### PR TITLE
simultaneous start, fixes #67

### LINKED ISSUE #67: Error when trying to run two new jobs simulatenously
The exception you get is:

```
Traceback (most recent call last):
  File "../crontabber/app.py", line 1286, in <module>
    local_main()
  File "../crontabber/app.py", line 1282, in local_main
    sys.exit(main(CronTabber))
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/generic_app.py", line 220, in main
    config_manager_cls
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/generic_app.py", line 329, in _do_main
    return_code = fix_exit_code(instance.main())
  File "../crontabber/app.py", line 774, in main
    self.run_all()
  File "../crontabber/app.py", line 934, in run_all
    self._run_one(job_class, class_config)
  File "../crontabber/app.py", line 956, in _run_one
    if not self.time_to_run(job_class, time_):
  File "../crontabber/app.py", line 1122, in time_to_run
    if next_run < utc_now():
TypeError: can't compare datetime.datetime to NoneType
```

I started with an absolutely clean database (no state row).
Configured 1 job that takes 60 seconds to complete. 
Execute this line in two separate terminals. One a couple of seconds after the other:

```
PYTHONPATH=. python ../crontabber/app.py --admin.conf=crontabber.ini
```

### PR DESCRIPTION
**Pull request by peterbe**
@twobraids r?
cc @rhelmer

### CODE REVIEW

**Comment by peterbe** [comment]
As discussed with @rhelmer 
We might still not be safe since there's a time between starting crontabber (loading the state, loading the configman classes, etc) and making that first `update crontabber set ongoing=now() where app_name=%`. 

Perhaps the next level solution is something in [this beast of a document](http://www.postgresql.org/docs/9.4/static/explicit-locking.html).

**Comment by twobraids** [comment]
Not knowing the full context,  I'm assuming two crontabber trying to start the same job at close to the same time. 

A starting place would be to first select for the row of the "crontabber"  table that is about to be updated with the "FOR UPDATE NOWAIT"  clause.  That'll automatically lock the row in a transaction safe manner.   Then any other process trying to execute the same "SELECT...  FOR UPDATE NOWAIT"  will fail with an error,  killing its enclosing transaction. One should ensure that the exception raised on failure is not considered retriable by the transaction executor. On encountering this exception, the second crontabber instance should deduce that the job it wanted to start is already going and it should move on to find the next non-dependent job.

**Comment by rhelmer** [comment]
@twobraids 

> Not knowing the full context, I'm assuming two crontabber trying to start the same job at close to the same time.

Right - the context is that we may get more than one node running crontabber simultaneously, due to the way we deploy in AWS (spin up new node, if it looks healthy then destroy the old.) Since it runs from crond every 5 minutes, we will almost certainly get multiple nodes starting crontabbers at the same second.

> A starting place would be to first select for the row of the "crontabber" table that is about to be updated with the "FOR UPDATE NOWAIT" clause. That'll automatically lock the row in a transaction safe manner. Then any other process trying to execute the same "SELECT... FOR UPDATE NOWAIT" will fail with an error, killing its enclosing transaction. One should ensure that the exception raised on failure is not considered retriable by the transaction executor. On encountering this exception, the second crontabber instance should deduce that the job it wanted to start is already going and it should move on to find the next non-dependent job.

This is what @selenamarie suggested as well - using row-level locking. We just need to be careful not to get into situations that could cause deadlock; since each crontabber only locks a single row at a time, I think we'll be ok, just something to consider.

**Comment by twobraids** [comment]
When using "... FOR UPDATE NOWAIT" I suspect that deadlock isn't even possible. Trying to acquire a lock and being unable to do so is an error that stops the transaction.  Deadlock can only happen if somebody waits for a lock.

**Comment by twobraids** [comment]
@peterbe so what is the status on this?

**Comment by peterbe** [comment]
I think we should review this and merge it. Its commit message is a bit poor since it's just a business level logic solution to the problem. A mere warm-up for the real solution maybe.

There is, however, a significant problem with the NOWAIT solution. If you have...:
```
$ ./crontabber.py --admin.conf=crontabber-tree1.ini &
$ ./crontabber.py --admin.conf=crontabber-tree2.ini
```
where the ini files define trees of jobs that are completely independent of each other. E.g. tree 1 might be a bunch of stored procedures and tree 2 is about cleaning up the disk and sending emails. 
Then we'd make it impossible to run these two separately. (unless you let the two ini files specify to use two different postgres servers)

The solution in my patch here is locked on a per job basis.

**Comment by twobraids** [comment]
the NO_WAIT solution is not a lock on the whole table, it is a lock only the single row that represents a job. The second crontabber fails iff the first crontabber has already started the very same job that the second crontabber is trying to start.  If they're running different jobs, there will be no failure and they can run independently.

As it stands now there is still a problem.  Both can start a transaction, and read the same data. The first tries to update and completes. The second tries to update and is blocked.  the first commits.  the second is unblocked and updates and commits.   the second did not encounter an error to stop it.  it only was blocked for the duration of the first's transaction.  the second makes its decision to continue based on the original read of the table, before the first commited its transaction.  the second has dirty data and doesn't know it.

File: crontabber/tests/test_crontabber.py [hunk 1]
```diff
@@ -2149,6 +2150,105 @@ def test_job_list_separated_sloppily(self):
             ok_('Ran FooJob' in infos, infos)
             ok_('Ran BarJob' in infos, infos)
 
+    def test_run_two_jobs_almost_simultaneously_first_time(self):
```
Comment (twobraids) [inline crontabber/tests/test_crontabber.py]: In order for this to really test if crontabber is trying to run them at them at the same time,  the transactions need to be guaranteed to overlap each other.  I suggest patching the transaction executor for both. It should hold one transaction open (before the update) until the other completes its select.

Comment (peterbe) [inline crontabber/tests/test_crontabber.py]: I'm not sure I understand. These two tests do test what the patch is fixing. I wrote the tests first, which failed consistently, then added the actual patch.

Comment (twobraids) [inline crontabber/tests/test_crontabber.py]: I believe our instance of Postgres is run in "read committed" isolation mode. That means that we don't do read locking on transactions, only write locking.  That means a connection may experience non-repeatable reads within a transaction (read a column from a row once and see it is 0,  read it again later within the transaction and find it has been changed to 1000). 

The test uses two threads each having its own database connection.  There is nothing in the threading or transaction models to prevent this order of operations between the threads: 
```
01 thread1 starts
02 thread2 starts
03 thread1 sleep 0.1
04 thread2 sleep 0.05
05 thread2 app.py:969 read job X transaction begins & ends
06 thread1 app.py:959 read job X transaction begins & ends
07 thread1 app.py:1135 read job X transaction begins & ends
08 thread2 app.py:1135 read job X transaction begins & ends
09 thread1 app.py:286 begin transaction
10 thread1 app.py:294 read job X
11 thread2 app.py:286 begin transaction
12 thread2 app.py:294 read job X
13 # both thread 1 & 2 have the same info about unstarted job X
14 thread2 app.py:357 updates job X (locking job X row)
15 thread1 app.py:357 updates job X (BLOCKED)
16 thread2 app.py:361 commits transaction assumes control of job X
17 thread1 app.py:357 unblocked - updates job X
18 thread1 app.py:361 commits transaction assumes control of job X
19 # both threads now think they own job X and are running it.
```

Is this an unlikely scenario?  All I can say is that my intuition says to me that this wide open opportunity for a concurrency problem.  The ultimate solution is to use the NOWAIT as it would have caused _line 15_ above to FAIL unstead of BLOCK.

Comment (selenamarie) [inline crontabber/tests/test_crontabber.py]: Yes, there is a concurrency problem. 

The right thing to do is to have every thread/process that's interested in *reading* and *using* the task information to get exclusive access. It is possible to change our transaction mode for crontabber to be SERIALIZABLE, but you'd still need to figure out how to handle the locking failures.

(The alternative (longer term) solution is to use a queue where the concurrency problem gets offloaded to the queue. I realize that's problematic because of the timing issues, but it is a feature that I think we might need in RelEng shortly.)

In any case, this blog post might be helpful in explaining the how of what needs to happen: http://binarybuffer.com/2012/12/grails-locking-rows-for-update-in-postgresql-with-nowait

It describes pretty much the same problem that you're addressing now.

This is another description of managing a queue in Postgres: http://www.depesz.com/2013/08/30/pick-a-task-to-work-on/

It's a good description of the problem as well, providing some alternative solutions. Depesz throws out the option of using NO WAIT because he doesn't want to manage the application-level logic (haha). I think *we* do want to manage this in the application.

**Comment by peterbe** [comment]
Those seem like awefully long blog posts :)

I think what already had me convinced was that @twobraids you can do the NOWAIT on a particular **row**. The problem we have potentially is that these two things might happen at the same time
```
(connection 1) UPDATE crontabber SET ongoing=NOW() WHERE app_name='foo';
(connection 2) UPDATE crontabber SET ongoing=NOW() WHERE app_name='foo';
```
And they're single-digit milliseconds apart. What might also happen, when it's a fresh new job is this:
```
(connection 1) SELECT * FROM crontabber WHERE app_name='foo';
(some python code concludes it's an empty recordset)
(connection 1) INSERT INTO crontabber (app_name, ongoing) VALUES('foo', NOW());
...same for connection 2...
```
What we need to do is to somehow say that any operation where `app_name='foo'` is locked to a single operation and any other connection trying to work with it should be blocked and error. 

I only yet to figure out how the heck to use this new fancy NOWAIT feature :)
...specifically for a column & value combo. 

If you can't acquire a lock for a specific column & value combo perhaps we simply ask postgres for a lock on the table for any update or insert. After all those above mentioned UPDATE and INSERT is what's critical that only one happens globally across multiple connections or transactions.

**Comment by selenamarie** [comment]
`SELECT FOR UPDATE NOWAIT` is associated with a predicate. So that just means that you simply add it where you would normally use a `SELECT` at the start of a transaction to find the row you're wanting to `UPDATE` later. Then, in the same transaction, `UPDATE` at the conclusion of the operation and `COMMIT` at the end.

**Comment by twobraids** [comment]
@peterbe, in otherwords, your scenario where the updates are single milliseconds apart, if you used the "...FOR UPDATE NOWAIT" the second of those updates will fail rather than block.  So there is no contention, there is a clear winner and clear loser.  To use the NOWAIT feature, simply read the row that you're going to want to update (or insert) with a "SELECT * from sometable FOR UPDATE NOWAIT".  Then do your update.

The second scenario that you mention with the inserts is also resolved with transactions.  Say two processes try to insert, the first will succeed and the second will BLOCK.  However, when the after the first commits, the second will unblock and immediately raise an exception.

There is no need in either of these scenarios to literally ask for a lock.  Transactions and properly worded SQL will imply the usage of locks automatically.

File: crontabber/tests/test_crontabber.py [hunk 2]
```diff
@@ -2149,6 +2150,170 @@ def test_job_list_separated_sloppily(self):
             ok_('Ran FooJob' in infos, infos)
             ok_('Ran BarJob' in infos, infos)
 
+    def test_run_two_jobs_simultaneously_first_time(self):
+        config_manager = self._setup_config_manager(
```
Comment (twobraids) [inline crontabber/tests/test_crontabber.py]: if you truncate the 'crontabber_log' table between these each of these next three tests, these tests pass travis...

**Comment by peterbe** [comment]
So after a day of deep debugging I think I've sort of figured it out. And I'm quite lost for a solution. 

I can get it to fail every time when I run this:
```
dropdb test_crontabber &&\
 createdb test_crontabber &&\
 nosetests crontabber/tests/test_crontabber.py:TestCrontabber.test_run_two_jobs_simultaneously_second_time \
 --with-cover --cover-erase --cover-package=crontabber
```
But if I remove the last part about coverage:
```
dropdb test_crontabber &&\
 createdb test_crontabber &&\
 nosetests crontabber/tests/test_crontabber.py:TestCrontabber.test_run_two_jobs_simultaneously_second_time
```
then it always succeeds. 

Basically, when you run without coverage, the tests runs so fast that the two attempts of...:
```
SELECT ... WHERE app_name='slow-job' FOR UPDATE NOWAIT;
UPDATE crontabber SET ongoing=... WHERE app_name='slow-job';
```
...clash. 

When you run **with coverage** it slows everything down and perhaps the coverage module hooks itself into the threading machinery. It slows it down so much that the lines above (the `SELECT ...NOWAIT; UPDATE`) happens milliseconds apart. 

Basically, I'm lost as to how to simulate that exact moment of two independent threads triggering something at the same time.

I'm right now tempted to put a pause on this as I'm starting to get more flustered and frustrated than I am productive.

**Comment by peterbe** [comment]
If I put a print of `time.time()` into the `__setitem__` method of the state class, then I can see that they are 0.01 seconds apart which is 10 milliseconds and I guess that's too wide a time-gap. 

When I remove the `--with-cover --cover-erase --cover-package=crontabber` the two, relevant, calls to `__setitem__()` are 0.00 seconds apart (the `time.time()`  module only reports in seconds with 2 significant figures) which is why they clash. 

Basically, when coverage is involved, how can we assert that two PG queries happen near simultaneously?

**Comment by peterbe** [comment]
Basically, we have a testing problem. I think the code is sane. I just don't know how to test it. Exceptions and possible `raise nose.SkipTest()`??? At your mercy for ideas @twobraids

**Comment by peterbe** [comment]
@rhelmer check it out! I'm using the `ongoing` value of the state if the NOWAIT doesn't block the second instance.

**Comment by peterbe** [comment]
Note-to-self: Once we land this, resolve https://bugzilla.mozilla.org/show_bug.cgi?id=1168957

File: crontabber/tests/test_crontabber.py (hunk 2 above)
Comment (peterbe) [inline crontabber/tests/test_crontabber.py]: For the record, that was not a valid idea. The problem was more profound than that.

**Comment by peterbe** [comment]
@twobraids I've addressed the concerns coveralls had with the reduced test coverage.

**Comment by adngdb** [comment]
@peterbe this needs to be rebased

**Comment by peterbe** [comment]
@adngdb rebased.

File: .travis.yml [hunk 3]
```diff
@@ -17,7 +17,7 @@ python:
     - 2.7
 
 addons:
-    postgresql: "9.4"
+    postgresql: "9.2"
```
Comment (adngdb) [inline .travis.yml]: I guess you can set it to 9.3 now, right? At least if you're trying to match what we use in Socorro.

File: crontabber/tests/test_crontabber.py [hunk 4]
```diff
@@ -330,6 +370,11 @@ def test_pop(self):
         popped = self.database.pop('foo', 'default')
         eq_(popped, 'default')
         assert_raises(KeyError, self.database.pop, 'bar')
+        try:
+            del self.database['bar']
+            assert False, "it should have raise a KeyError on del"
+        except KeyError:
+            pass
```
Comment (adngdb) [inline crontabber/tests/test_crontabber.py]: Why not use ``assert_raises`` here? 

```python
assert_raises(KeyError, lambda k: del self.database[k], 'bar')
```

File: exampleapp/foo.py [hunk 5]
```diff
@@ -6,5 +6,9 @@ class FooCronApp(BaseCronApp):
     app_name = 'foo'
 
     def run(self):
+        # from time import sleep
+        # print "Starting to sleep..."
+        # sleep(60)
+        # print "Done sleepin."
```
Comment (adngdb) [inline exampleapp/foo.py]: debug leftover?

**Comment by adngdb** [comment]
OK, I think I understand ``FOR UPDATE NOWAIT``, and this code looks good to me. The tests seems solid too. 

You did mention problems with the tests though, did you figure those out? If not, do you want to talk about it next time we meet?

File: .travis.yml (hunk 3 above)
Comment (peterbe) [inline .travis.yml]: If only I can remember **why** I changed this to 9.2 for this PR. Grr.... Time sucks.

File: crontabber/tests/test_crontabber.py (hunk 4 above)
Comment (peterbe) [inline crontabber/tests/test_crontabber.py]: excellent idea

Comment (peterbe) [inline crontabber/tests/test_crontabber.py]: oh, now I remember. `SyntaxError` on using a del inside a lambda.

File: exampleapp/foo.py (hunk 5 above)
Comment (peterbe) [inline exampleapp/foo.py]: No. Good to have when debugging the exampleapp.

File: .travis.yml (hunk 3 above)
Comment (peterbe) [inline .travis.yml]: Let's try with 9.4 and see what happens.

File: crontabber/tests/test_crontabber.py (hunk 4 above)
Comment (willkg) [inline crontabber/tests/test_crontabber.py]: These are unitttest TestCase dervied, right? If so, you could use the context manager.

https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertRaises

I don't remember if nose facilitates kind of thing or not. pytest has this:

http://pytest.org/latest/assert.html#assertions-about-expected-exceptions

Maybe nose has something similar.

Comment (peterbe) [inline crontabber/tests/test_crontabber.py]: `assert_raises` and `self.assertRaises` are similar. The second argument is a callable. In my case `del self.database['bar']` is not a callable.

Comment (willkg) [inline crontabber/tests/test_crontabber.py]: No... they're not similar. I'm pretty sure `assert_raises` can't work as a context manager. I don't know if nose has something else that does work as a context manager. It's definitely worth switching to a context manager form.

```
with self.assertRaises(KeyError):
    del self.database['bar']
```

Comment (willkg) [inline crontabber/tests/test_crontabber.py]: Oh, I might take it back. Looks like nose pep8-ifies unittest testcase things. So you might also be able to do:

```
with assert_raises(KeyError):
    del self.database['bar']
```

Definitely worth trying.

Comment (peterbe) [inline crontabber/tests/test_crontabber.py]: Cool! it worked!

**Comment by peterbe** [comment]
> You did mention problems with the tests though, did you figure those out? If not, do you want to talk about it next time we meet? 

So there's a complicated issued. It's extremely hard to start two crontabber apps (instances of the `Crontabber` with apps configured) at the same time in unit tests. I was able to do it by spawning threads and I was able to do test-driven development of this low-level protection of simultaneous SQL insert/update attempts. HOWEVER, when I run the tests with coverage (which we do in travis always) that coverage injection code always adds a tiny amount of inertia. So much so that it's near impossible to make those spawned threads hit the same SQL update/insert at the "same time". So, the reason this PR is so complicated is because I had to wrangle myself around with other mocking and stuff so that it pushes the code where it matters. 

Remember, this PR is about TWO different "race condition" cases. 
* The under 10-20 ms race condition. Where to independent python processors try to write to the same global postgres database at the same "exact" time. That's what the "FOR UPDATE WAIT" stuff is for. 
* The seconds and minutes race condition. Today, in socorro we use bash locking to not trigger crontabber_app whilst it's still running because crontabber assumed you won't. Now, as of this PR, when it's about to start a crontabber it will plainly look into the state (which is a row in Postgres) to see if the job has a `ongoing` value that is not null. This means that if you have, on your laptop, two terminals prepared to start the same job, when you switch to the second terminal (takes your fingers about 1-2 seconds) you'll get an error on the second terminal. This means that we can stop using bash locking (no reason to though) and let the crontab just fire every 5 minutes and if it's busy it simply errors out and tries again in 5 minutes. 

One important consequence with this PR is that if you start a crontabber job that takes, say, 60 seconds and during that time you pull the plug on the postgres database/connection/harddrive crontabber doesn't get a chance to say it's not longer ongoing when the job has finished. That means you can have a job "stuck" as ongoing "forever". We could put in a cap on that. [Here](https://github.com/mozilla/crontabber/pull/69/files#diff-2aad6bd5c1f660ea5b7d4d4a97c898d5R333). So we say the job can only be ongoing for a max of, say 2 hours. That means that if it'd deadlock (due to postgres's plug being pulled) prevents new jobs it'd simply bulldozer over that.

**Comment by peterbe** [comment]
So, I'm glad I looked. If you run a second instance (in a different terminal, whilst a job is ongoing) you get a nasty error:
```
...
2017-03-01 11:00:36,280 DEBUG - MainThread -  - MainThread - about to run <class 'foo.FooCronApp'>
2017-03-01 11:00:36,282 ERROR - MainThread -  - MainThread - Exception raised during crontabber.connection_factory transaction
Traceback (most recent call last):
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/transaction_executor.py", line 46, in __call__
    result = function(connection, *args, **kwargs)
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/app.py", line 356, in __setitem__
    raise RowLevelLockError('actively ongoing')
RowLevelLockError: actively ongoing
2017-03-01 11:00:36,283 DEBUG - MainThread -  - MainThread - error when running <class 'foo.FooCronApp'> on None
Traceback (most recent call last):
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/app.py", line 1030, in _run_one
    for last_success in self._run_job(job_class, config, info):
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/app.py", line 1202, in _run_job
    self._set_ongoing_job(class_)
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/app.py", line 1227, in _set_ongoing_job
    self.job_state_database[app_name] = info
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/app.py", line 93, in _do_transaction
    **kwargs
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/transaction_executor.py", line 46, in __call__
    result = function(connection, *args, **kwargs)
  File "/Users/peterbe/dev/PYTHON/crontabber/crontabber/app.py", line 356, in __setitem__
    raise RowLevelLockError('actively ongoing')
RowLevelLockError: actively ongoing
2017-03-01 11:00:36,294 INFO - MainThread -  - MainThread - done.
```
Also, the exit code is `0` when it should be anything but.
